cmd/gardener-external-admission-controller/main.go

```diff
@@ -0,0 +1,211 @@
+	deserializer := serializer.NewCodecFactory(scheme).UniversalDeserializer()
+
+	mux := http.NewServeMux()
+
+	pvInitializerHandler := &pvInitializerHandler{deserializer}
+	mux.HandleFunc("/webhooks/mutate-pv-initializers", pvInitializerHandler.mutate)
+
+	srv := &http.Server{
+		Addr:    fmt.Sprintf("%s:%d", bindAddress, port),
+		Handler: mux,
+	}
+
+	go func() {
+		logger.Infof("Starting Gardener external admission controller...")
+		if err := srv.ListenAndServeTLS(certPath, keyPath); err != http.ErrServerClosed {
+			logger.Errorf("Could not start HTTPS server: %v", err)
+		}
+	}()
+
+	<-stopCh
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	if err := srv.Shutdown(ctx); err != nil {
+		logger.Errorf("Error when shutting down HTTPS server: %v", err)
+	}
+	logger.Info("HTTPS servers stopped.")
+}
+
+type pvInitializerHandler struct {
+	codecs runtime.Decoder
+}
+
+func (p *pvInitializerHandler) mutate(w http.ResponseWriter, r *http.Request) {
+	var (
+		body           []byte
+		receivedReview = admissionv1beta1.AdmissionReview{}
+	)
+
+	// Read HTTP request body into variable.
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+
+	// Verify that the correct content-type header has been sent.
+	if contentType := r.Header.Get("Content-Type"); contentType != "application/json" {
+		err := fmt.Errorf("contentType=%s, expect %s", contentType, "application/json")
+		logger.Errorf(err.Error())
+		respond(w, &admissionv1beta1.AdmissionResponse{Result: &metav1.Status{Message: err.Error()}})
+		return
+	}
+
+	// Deserialize HTTP request body into admissionv1beta1.AdmissionReview object.
+	if _, _, err := p.codecs.Decode(body, nil, &receivedReview); err != nil {
+		logger.Errorf(err.Error())
+		respond(w, &admissionv1beta1.AdmissionResponse{Result: &metav1.Status{Message: err.Error()}})
+		return
+	}
+
+	// If the request field is empty then do not admit (invalid body).
+	if receivedReview.Request == nil {
+		err := fmt.Errorf("invalid request body (missing admission request)")
+		logger.Errorf(err.Error())
+		respond(w, &admissionv1beta1.AdmissionResponse{Result: &metav1.Status{Message: err.Error()}})
+		return
+	}
+	// If the request does not indicate the correct operation (CREATE,UPDATE) we allow the review without further doing.
+	if receivedReview.Request.Operation != admissionv1beta1.Create && receivedReview.Request.Operation != admissionv1beta1.Update {
+		respond(w, &admissionv1beta1.AdmissionResponse{})
+		return
+	}
+
+	// Check that resource is a persistentvolume
+	persistentVolumeResource := metav1.GroupVersionResource{Group: "", Version: "v1", Resource: "persistentvolumes"}
+	if receivedReview.Request.Resource != persistentVolumeResource {
+		err := fmt.Errorf("invalid object (no persistentvolume)")
+		logger.Errorf(err.Error())
+		respond(w, &admissionv1beta1.AdmissionResponse{Result: &metav1.Status{Message: err.Error()}})
+		return
+	}
```

Comment:
Maybe we can factor out all the validation/verification code into another function ?